landoapi/hgexports.py

```diff
@@ -135,54 +217,153 @@ def commit_description(self) -> bytes:
+
+    def get_timestamp(self) -> str:
+        """Return an `hg export` formatted timestamp."""
+        date = self.get_header("Date")
+        if not date:
+            raise ValueError("Could not determine patch timestamp from header.")
+
+        return get_timestamp_from_hg_date_header(date)
+
+
+class GitPatchHelper(PatchHelper):
+    """Helper class for parsing Mercurial patches/exports."""
+
+    def __init__(self, fileobj: io.BytesIO):
+        super().__init__(fileobj)
+        self.parse_patch()
+
+    def get_header(self, name: bytes | str) -> Optional[str]:
+        """Get the headers from the message."""
+        name = name.decode("utf-8") if isinstance(name, bytes) else name
+
+        # `EmailMessage` will return `None` if the header isn't found.
+        return self.message[name]
+
+    def verify_content(self, content: str) -> str:
+        """Verify the diff is present in the patch."""
+        subject_header = self.get_header("Subject")
+        if not subject_header:
+            raise ValueError("No valid subject header for commit message.")
+
+        # Start the commit message from the stripped subject line.
+        commit_message_lines = [
+            subject_header.removeprefix("[PATCH] ").removesuffix("\n")
+        ]
+
+        # Create an iterator for the lines of the patch.
+        line_iterator = iter(content.splitlines())
+
+        # Add each line to the commit message until we hit `---`.
+        for i, line in enumerate(line_iterator):
+            if line == "---":
+                break
+
+            # Add a newline after the subject line if this is a multi-line
+            # commit message.
+            if i == 0:
+                commit_message_lines += [""]
+
+            commit_message_lines.append(line)
+
+        self.commit_message = "\n".join(commit_message_lines)
+
+        # Move through the patch until we find the start of the diff.
+        # Add the diff start line to the diff.
+        diff_lines = []
+        for line in line_iterator:
+            if GitPatchHelper._is_diff_line(line):
+                diff_lines.append(line)
+                break
+        else:
+            raise ValueError("Patch is malformed, could not find start of patch diff.")
+
+        # The diff is the remainder of the patch, except the last two lines of Git version info.
+        remaining_lines = list(line_iterator)
+        diff_lines += list(remaining_lines[:-2])
+        return "\n".join(diff_lines)
+
+    def parse_patch(self):
+        """Parse the Git patch file for headers and diff content."""
+        self.message = email.message_from_bytes(
+            self.patch.read(), policy=default_email_policy
+        )
+        self.diff = self.verify_content(self.message.get_content())
+
+    def commit_description(self) -> str:
+        """Returns the commit description."""
+        return self.commit_message
+
+    def get_diff(self) -> str:
+        """Return the patch diff."""
+        return self.diff
```

Comment:
What is the point of these methods? If it's to make naming more consistent, perhaps `commit_description` should be `get_commit_description`.

Reply:
`commit_description` is carried over from the existing `PatchHelper` code.

`get_diff` has been added to retrieve the diff from the patch since we need it for adding to the landing job. The existing `PatchHelper` code had a `write_diff` method which I refactored to instead return the diff, and made `write_diff` just write the value returned from `get_diff`.